builder/azure/arm/builder.go

```diff
@@ -281,7 +281,7 @@ func (b *Builder) Run(ctx context.Context, ui packersdk.Ui, hook packersdk.Hook)
 			NewStepGetCertificate(azureClient, ui),
 			NewStepSetCertificate(&b.config, ui),
 			NewStepValidateTemplate(azureClient, ui, &b.config, deploymentName, GetVirtualMachineDeployment),
-			NewStepDeployTemplate(azureClient, ui, &b.config, deploymentName, GetVirtualMachineDeployment),
+			NewStepDeployTemplate(azureClient, ui, &b.config, deploymentName, GetVirtualMachineDeployment, true),
```

Comment:
Another option, which I think helps readability, is to add an enum the to the step_deploy_template that can act as the last parameter, which will also help make the StepDeployTemplate step easier to grok, without having to look into its implementation.


```go
type DeploymentTemplateType int
const(
  VirtualMachineTemplate DeploymentTemplateType = iota
  KeyVaultTemplate
)
```

Then you can change the function to take in that type

```go
func NewStepDeployTemplate(client *AzureClient, ui packersdk.Ui, config *Config, deploymentName string, factory templateFactoryFunc, templateType DeploymentTemplateType ) *StepDeployTemplate {
	var step = &StepDeployTemplate{
		client:                     client,
		say:                        func(message string) { ui.Say(message) },
		error:                      func(e error) { ui.Error(e.Error()) },
		config:                     config,
		factory:                    factory,
		name:                       deploymentName,
		deploymentType:             templateType,
	}
```

```go
func (s *StepDeployTemplate) Cleanup(state multistep.StateBag){

//redacted
if s.deploymentTyp == KeyVaultTemplate {
	ui.Say("\nDeleting KeyVault created during build")
		err := s.delete(context.TODO(), deploymentName, resourceGroupName)
		if err != nil {
			s.say("Failed to delete the managed Additional Disk!")
			s.reportIfError(err, resourceGroupName)
		}
        return
}
// otherwise handle the VirtualMachine deletion case.

Reply:
Implemented this suggestion, thanks for the feedback Wilken!